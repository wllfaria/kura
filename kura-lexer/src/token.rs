#[allow(clippy::module_inception)]
mod token;

mod kind;
mod operator;
mod primitive;
mod value;

pub use kind::Kind;
pub use operator::Operator;
pub use primitive::{FloatSizes, IntSizes, Primitive, UIntSizes};
pub use token::{IntoToken, Location, Token};
pub use value::Value;
