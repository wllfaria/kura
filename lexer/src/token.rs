#[derive(Debug)]
pub enum NumSize {
    F8,
    F16,
    F32,
    F64,
    I8,
    I16,
    I32,
    I64,
    U8,
    U16,
    U32,
    U64,
    Usize,
    Isize,
}

impl std::fmt::Display for NumSize {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            NumSize::F8 => write!(f, "f8"),
            NumSize::F16 => write!(f, "f16"),
            NumSize::F32 => write!(f, "f32"),
            NumSize::F64 => write!(f, "f64"),
            NumSize::I8 => write!(f, "i8"),
            NumSize::I16 => write!(f, "i16"),
            NumSize::I32 => write!(f, "i32"),
            NumSize::I64 => write!(f, "i64"),
            NumSize::U8 => write!(f, "u8"),
            NumSize::U16 => write!(f, "u16"),
            NumSize::U32 => write!(f, "u32"),
            NumSize::U64 => write!(f, "u64"),
            NumSize::Usize => write!(f, "usize"),
            NumSize::Isize => write!(f, "isize"),
        }
    }
}

impl TryFrom<&str> for NumSize {
    type Error = ();

    fn try_from(value: &str) -> Result<Self, Self::Error> {
        match value {
            "i8" => Ok(NumSize::I8),
            "i16" => Ok(NumSize::I16),
            "i32" => Ok(NumSize::I32),
            "i64" => Ok(NumSize::I64),
            "u8" => Ok(NumSize::U8),
            "u16" => Ok(NumSize::U16),
            "u32" => Ok(NumSize::U32),
            "u64" => Ok(NumSize::U64),
            "f8" => Ok(NumSize::F8),
            "f16" => Ok(NumSize::F16),
            "f32" => Ok(NumSize::F32),
            "f64" => Ok(NumSize::F64),
            "usize" => Ok(NumSize::Usize),
            "isize" => Ok(NumSize::Isize),
            _ => Err(()),
        }
    }
}

#[derive(Debug)]
pub enum Token<'tok> {
    Integer { value: u64, size: Option<NumSize> },
    SignedInteger { value: i64, size: Option<NumSize> },
    Float { value: f64, size: Option<NumSize> },
    Minus,
    String(&'tok str),
}

impl std::fmt::Display for Token<'_> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Token::Integer { value, size } => write!(
                f,
                "{}{}",
                value,
                size.as_ref().map(|s| s.to_string()).unwrap_or_default()
            ),
            Token::Float { value, size } => write!(
                f,
                "{}{}",
                value,
                size.as_ref().map(|s| s.to_string()).unwrap_or_default()
            ),
            Token::SignedInteger { value, size } => write!(
                f,
                "{}{}",
                value,
                size.as_ref().map(|s| s.to_string()).unwrap_or_default()
            ),
            Token::Minus => write!(f, "-"),
            Token::String(s) => write!(f, "{s}"),
        }
    }
}
