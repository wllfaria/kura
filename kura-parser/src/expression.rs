use kura_lexer::token::{Kind, Location, Operator, Primitive, Token, Value};
use kura_lexer::{Lexer, TransposeRef};

use crate::Expression;

mod precedences {
    pub const BASE: u8 = 0;
    pub const SUM: u8 = 3;
    pub const MUL: u8 = 4;
    pub const ASSOC: u8 = 5;
    pub const APPLY: u8 = 6;
}

fn get_precedence(operator: Operator) -> u8 {
    match operator {
        Operator::Plus | Operator::Minus => precedences::SUM,
        Operator::Star | Operator::Slash => precedences::MUL,
        Operator::And => precedences::ASSOC,
        Operator::LeftParen | Operator::EqualEqual | Operator::NotEqual => precedences::APPLY,
        _ => precedences::BASE,
    }
}

pub fn parse_expression<'parser>(lexer: &mut Lexer<'parser>, base_expr: bool) -> Result<Expression<'parser>, String> {
    match lexer.peek().transpose().map_err(ToString::to_string)? {
        Some(token) => match token.kind {
            Kind::Var | Kind::Const => parse_variable(lexer),
            _ => parse_with_precedence(lexer, precedences::BASE, base_expr),
        },
        None => unreachable!(),
    }
}

pub fn parse_identifier<'parser>(lexer: &mut Lexer<'parser>) -> Result<(Expression<'parser>, &'parser str), String> {
    let name_and_loc = lexer
        .next()
        .transpose()
        .map_err(|e| e.to_string())?
        .map(|token| match token.kind {
            Kind::Value(Value::Ident(name)) => (name, token.location),
            _ => ("", token.location),
        });

    match name_and_loc {
        Some(("", location)) => Err(location.to_string()),
        Some((name, location)) => Ok((Expression::Ident { name, location }, name)),
        None => Err("TODO".into()),
    }
}

fn parse_expr_block<'parser>(lexer: &mut Lexer<'parser>, base_expr: bool) -> Result<Expression<'parser>, String> {
    let mut expressions = vec![];

    let block_start = lexer.expect(Kind::Op(Operator::LeftBrace)).map_err(|e| e.to_string())?;

    loop {
        match lexer.peek().transpose().map_err(|e| e.to_string())? {
            Some(token) if matches!(token.kind, Kind::Op(Operator::RightBrace)) => {
                break;
            }
            None => break,
            _ => (),
        }

        let expr = parse_expression(lexer, base_expr)?;
        expressions.push(expr);
    }

    let block_end = lexer
        .expect(Kind::Op(Operator::RightBrace))
        .map_err(|e| e.to_string())?;

    let location = block_start.location.start_byte..block_end.location.end_byte;
    Ok(Expression::Block {
        expressions,
        location: location.into(),
    })
}

fn parse_variable<'parser>(lexer: &mut Lexer<'parser>) -> Result<Expression<'parser>, String> {
    let keyword = lexer
        .expect_one_of(&[Kind::Var, Kind::Const])
        .map_err(|e| e.to_string())?;
    let mutable = matches!(keyword.kind, Kind::Var);

    let (_, name) = parse_identifier(lexer)?;

    let typ = match lexer.peek().transpose().map_err(|e| e.to_string())? {
        Some(token) if matches!(token.kind, Kind::Op(Operator::Colon)) => {
            lexer.next().transpose().map_err(|e| e.to_string())?;
            Some(parse_identifier(lexer)?)
        }
        _ => None,
    };

    lexer.expect(Kind::Op(Operator::Equal)).map_err(|e| e.to_string())?;

    let value = match lexer.peek().transpose().map_err(|e| e.to_string())? {
        Some(token) => match token.kind {
            Kind::Op(Operator::LeftBrace) => parse_expr_block(lexer, false)?,
            _ => parse_expression(lexer, false)?,
        },
        _ => unreachable!(),
    };

    lexer.expect(Kind::Op(Operator::SemiColon)).map_err(|e| e.to_string())?;

    let location = Location::new(keyword.location.start_byte, value.location().end_byte);
    Ok(Expression::Var {
        mutable,
        typ: typ.map(|(typ, _)| Box::new(typ)),
        name,
        value: Box::new(value),
        location,
    })
}

fn parse_if_expression<'parser>(lexer: &mut Lexer<'parser>) -> Result<Expression<'parser>, String> {
    let keyword = lexer.expect(Kind::If).map_err(|e| e.to_string())?;

    let condition = parse_expression(lexer, false)?;

    let body = parse_expr_block(lexer, true)?;

    let mut falsy_branches = vec![];

    while let Some(keyword) = lexer.peek().transpose().map_err(|e| e.to_string())? {
        if keyword.kind == Kind::Else {
            // consume `else`
            lexer.next().transpose().map_err(|e| e.to_string())?;

            if lexer
                .peek()
                .transpose()
                .map_err(|e| e.to_string())?
                .map(|t| matches!(t.kind, Kind::If))
                .unwrap_or(false)
            {
                let else_if = parse_if_expression(lexer)?;
                falsy_branches.push(else_if);
            } else {
                let else_block = parse_expr_block(lexer, true)?;
                falsy_branches.push(else_block);
                break;
            }
        } else {
            break;
        }
    }

    let end = match falsy_branches.last() {
        Some(branch) => branch.location().end_byte,
        None => body.location().end_byte,
    };

    let location = keyword.location.start_byte..end;
    Ok(Expression::If {
        condition: Box::new(condition),
        truthy: Box::new(body),
        falsy: falsy_branches,
        location: location.into(),
    })
}

fn parse_value<'parser>(lexer: &mut Lexer<'parser>) -> Result<Expression<'parser>, String> {
    let Some(Token { kind, .. }) = lexer.peek().transpose().map_err(|e| e.to_string())? else {
        unreachable!();
    };

    let Kind::Value(value) = kind else {
        unreachable!();
    };

    match value {
        Value::Primitive(_) => parse_primitive(lexer),
        Value::Ident(_) => Ok(parse_identifier(lexer)?.0),
        t => todo!("{t}"),
    }
}

fn parse_operation<'parser>(lexer: &mut Lexer<'parser>) -> Result<Expression<'parser>, String> {
    let Some(Token { kind, .. }) = lexer.peek().transpose().map_err(|e| e.to_string())? else {
        unreachable!();
    };

    let Kind::Op(op) = kind else {
        unreachable!();
    };

    match op {
        Operator::LeftParen => {
            lexer.next().transpose().map_err(|e| e.to_string())?;
            let left = parse_with_precedence(lexer, precedences::BASE, false)?;
            lexer
                .expect(Kind::Op(Operator::RightParen))
                .map_err(|e| e.to_string())?;
            Ok(left)
        }
        Operator::LeftBrace => parse_expr_block(lexer, true),
        t => todo!("{t:?}"),
    }
}

fn parse_fun_call<'parser>(
    lexer: &mut Lexer<'parser>,
    ident: Expression<'parser>,
    base_expr: bool,
) -> Result<Expression<'parser>, String> {
    lexer.expect(Kind::Op(Operator::LeftParen)).map_err(|e| e.to_string())?;

    let mut arguments = vec![];

    loop {
        match lexer.peek().transpose().map_err(|e| e.to_string())? {
            Some(token) if matches!(token.kind, Kind::Op(Operator::RightParen)) => {
                break;
            }
            Some(token) if matches!(token.kind, Kind::Op(Operator::Comma)) => {
                lexer.next().transpose().map_err(|e| e.to_string())?;
                continue;
            }
            None => break,
            _ => (),
        }

        let arg = parse_expression(lexer, false)?;

        arguments.push(arg);
    }

    let close_paren = lexer
        .expect(Kind::Op(Operator::RightParen))
        .map_err(|e| e.to_string())?;

    if base_expr {
        lexer.expect(Kind::Op(Operator::SemiColon)).map_err(|e| e.to_string())?;
    }

    let location = ident.location().start_byte..close_paren.location.end_byte;
    let expr = Expression::FunCall {
        ident: Box::new(ident),
        location: location.into(),
        arguments,
    };
    Ok(expr)
}

fn parse_assign<'parser>(lexer: &mut Lexer<'parser>, left: Expression<'parser>) -> Result<Expression<'parser>, String> {
    lexer.expect(Kind::Op(Operator::Equal)).map_err(|e| e.to_string())?;

    let value = match lexer.peek().transpose().map_err(|e| e.to_string())? {
        Some(token) if matches!(token.kind, Kind::Op(Operator::LeftBrace)) => parse_expr_block(lexer, false)?,
        Some(_) => parse_expression(lexer, false)?,
        _ => unreachable!(),
    };

    let closing = lexer.expect(Kind::Op(Operator::SemiColon)).map_err(|e| e.to_string())?;

    let location = left.location().start_byte..closing.location.end_byte;

    Ok(Expression::Assign {
        ident: Box::new(left),
        location: location.into(),
        value: Box::new(value),
    })
}

fn parse_with_precedence<'parser>(
    lexer: &mut Lexer<'parser>,
    min_precedence: u8,
    base_expr: bool,
) -> Result<Expression<'parser>, String> {
    let mut left = match lexer.peek().transpose().map_err(|e| e.to_string())? {
        Some(token) => match &token.kind {
            Kind::Value(_) => parse_value(lexer)?,
            Kind::Op(_) => parse_operation(lexer)?,
            Kind::Return => parse_return_expression(lexer)?,
            Kind::If => parse_if_expression(lexer)?,
            t => todo!("{t:?}"),
        },
        None => todo!(),
    };

    if let Expression::Ident { .. } = left {
        match lexer.peek().transpose().map_err(|e| e.to_string())? {
            Some(token) if matches!(token.kind, Kind::Op(Operator::LeftParen)) => {
                return parse_fun_call(lexer, left, base_expr);
            }
            Some(token) if matches!(token.kind, Kind::Op(Operator::Equal)) => return parse_assign(lexer, left),
            _ => (),
        }
    }

    loop {
        let Some(next) = lexer.peek().transpose().map_err(|e| e.to_string())? else {
            return Ok(left);
        };

        let Kind::Op(operator) = next.kind else {
            return Ok(left);
        };

        if !next.kind.is_binary_op() {
            return Ok(left);
        }

        let precedence = get_precedence(operator);

        if precedence <= min_precedence {
            break;
        }

        let Some(_) = lexer.next().transpose().map_err(|e| e.to_string())? else {
            unreachable!();
        };

        let right = parse_with_precedence(lexer, precedence, false)?;

        let location = Location::new(left.location().start_byte, right.location().end_byte);
        left = Expression::BinaryOp {
            lhs: Box::new(left),
            operator,
            rhs: Box::new(right),
            location,
        };
    }

    Ok(left)
}

fn parse_return_expression<'parser>(lexer: &mut Lexer<'parser>) -> Result<Expression<'parser>, String> {
    let keyword = lexer.expect(Kind::Return).map_err(|e| e.to_string())?;

    let value = parse_expression(lexer, false)?;

    let ending_semi = lexer.expect(Kind::Op(Operator::SemiColon)).map_err(|e| e.to_string())?;

    let location = keyword.location.start_byte..ending_semi.location.end_byte;
    Ok(Expression::Return {
        value: Box::new(value),
        location: location.into(),
    })
}

fn parse_primitive<'parser>(lexer: &mut Lexer<'parser>) -> Result<Expression<'parser>, String> {
    let (primitive, location) = match lexer.next().transpose().map_err(|e| e.to_string())? {
        Some(Token {
            kind: Kind::Value(Value::Primitive(primitive)),
            location,
            ..
        }) => (primitive, location),
        _ => unreachable!(),
    };

    match primitive {
        Primitive::Int { value, size } => Ok(Expression::IntLiteral { value, size, location }),
        Primitive::UInt { value, size } => Ok(Expression::UintLiteral { value, size, location }),
        Primitive::Float { value, size } => Ok(Expression::FloatLiteral { value, size, location }),
        Primitive::Bool(value) => Ok(Expression::Bool { value, location }),
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::Parser;

    fn make_sut(source: &str) -> Parser<'_> {
        let lexer = Lexer::new(source);
        Parser::new(source, lexer)
    }

    #[test]
    fn simple_math() {
        let simple_math_expr = "1 + 2 * 3 - 4";

        let mut parser = make_sut(simple_math_expr);

        let math_expr_ast = match parse_expression(&mut parser.lexer, true) {
            Ok(expr) => expr,
            Err(e) => panic!("{e:?}"),
        };

        insta::assert_debug_snapshot!(math_expr_ast);
    }

    #[test]
    fn mutable_variable() {
        let variables = "var hello = 1 + 2 * 3;";
        let mut parser = make_sut(variables);

        let variables_ast = match parse_expression(&mut parser.lexer, true) {
            Ok(expr) => expr,
            Err(e) => panic!("{e:?}"),
        };

        insta::assert_debug_snapshot!(variables_ast);
    }

    #[test]
    fn immutable_constant() {
        let variables = "const hello = 1 + 2 * 3;";
        let mut parser = make_sut(variables);

        let variables_ast = match parse_expression(&mut parser.lexer, true) {
            Ok(expr) => expr,
            Err(e) => panic!("{e:?}"),
        };

        insta::assert_debug_snapshot!(variables_ast);
    }

    #[test]
    fn if_statement() {
        let source = r#"
            if something == another_thing && 1 != 2 {
                const this_is_a_var = 10 + 3;
            } else if something == 10 {
                const this_is_another = 10 + 10;
            } else {
                const omg = 1 + 1;
            }
        "#;
        let mut parser = make_sut(source);

        let if_ast = match parse_expression(&mut parser.lexer, true) {
            Ok(expr) => expr,
            Err(e) => panic!("{e:?}"),
        };

        insta::assert_debug_snapshot!(if_ast);
    }

    #[test]
    fn if_as_variable_value() {
        let source = r#"
            const my_var = if something == another_thing {
                10 + 10
            } else {
                20 + 20
            };
        "#;

        let mut parser = make_sut(source);
        let let_if_ast = match parse_expression(&mut parser.lexer, true) {
            Ok(expr) => expr,
            Err(e) => panic!("{e:?}"),
        };

        insta::assert_debug_snapshot!(let_if_ast);
    }
}
