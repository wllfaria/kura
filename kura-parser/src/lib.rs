mod expression;

use kura_lexer::token::{FloatSizes, IntSizes, Kind, Location, Operator, Token, UIntSizes};
use kura_lexer::{Lexer, TransposeRef};

use crate::expression::{parse_expression, parse_identifier};

#[derive(Debug)]
pub enum Statement<'ast> {
    FunArgument {
        name: &'ast str,
        arg_type: Box<Expression<'ast>>,
        location: Location,
    },
    Fun {
        name: &'ast str,
        arguments: Vec<Statement<'ast>>,
        body: Vec<Expression<'ast>>,
        return_type: Option<Box<Expression<'ast>>>,
        location: Location,
    },
}

#[derive(Debug)]
pub enum Expression<'ast> {
    Var {
        mutable: bool,
        name: &'ast str,
        typ: Option<Box<Expression<'ast>>>,
        value: Box<Expression<'ast>>,
        location: Location,
    },
    Bool {
        value: bool,
        location: Location,
    },
    If {
        condition: Box<Expression<'ast>>,
        location: Location,
        truthy: Box<Expression<'ast>>,
        // each `else` or `else if` are added here, if nothing, then
        // this will be an empty vec
        //
        // TODO: maybe use Options so we dont allocate a vector
        falsy: Vec<Expression<'ast>>,
    },
    FunCall {
        ident: Box<Expression<'ast>>,
        location: Location,
        arguments: Vec<Expression<'ast>>,
    },
    Assign {
        ident: Box<Expression<'ast>>,
        location: Location,
        value: Box<Expression<'ast>>,
    },
    Ident {
        name: &'ast str,
        location: Location,
    },
    Block {
        expressions: Vec<Expression<'ast>>,
        location: Location,
    },
    UintLiteral {
        value: u64,
        size: Option<UIntSizes>,
        location: Location,
    },
    FloatLiteral {
        value: f64,
        size: Option<FloatSizes>,
        location: Location,
    },
    IntLiteral {
        value: i64,
        size: Option<IntSizes>,
        location: Location,
    },
    BinaryOp {
        operator: Operator,
        lhs: Box<Expression<'ast>>,
        rhs: Box<Expression<'ast>>,
        location: Location,
    },
    Return {
        value: Box<Expression<'ast>>,
        location: Location,
    },
}

impl Expression<'_> {
    fn location(&self) -> Location {
        match self {
            Expression::Var { location, .. } => *location,
            Expression::If { location, .. } => *location,
            Expression::Ident { location, .. } => *location,
            Expression::Bool { location, .. } => *location,
            Expression::Block { location, .. } => *location,
            Expression::FunCall { location, .. } => *location,
            Expression::Assign { location, .. } => *location,
            Expression::UintLiteral { location, .. } => *location,
            Expression::Return { location, .. } => *location,
            Expression::FloatLiteral { location, .. } => *location,
            Expression::IntLiteral { location, .. } => *location,
            Expression::BinaryOp { location, .. } => *location,
        }
    }
}

pub struct Parser<'par> {
    pub source: &'par str,
    pub lexer: Lexer<'par>,
}

impl<'par> Parser<'par> {
    pub fn new(source: &'par str, lexer: Lexer<'par>) -> Self {
        Self { source, lexer }
    }

    pub fn parse(mut self) -> Result<Vec<Statement<'par>>, String> {
        let mut statements = vec![];

        while !self.lexer.is_empty() {
            let statement = self.parse_statement()?;
            statements.push(statement);
        }

        Ok(statements)
    }

    fn parse_statement(&mut self) -> Result<Statement<'par>, String> {
        match self.lexer.peek().transpose().map_err(|e| e.to_string())? {
            Some(token) => match &token.kind {
                Kind::Fun => self.parse_function(),
                t => todo!("{t:?}"),
            },
            _ => todo!(),
        }
    }

    fn parse_function_args(&mut self, _: &Token<'_>) -> Result<Vec<Statement<'par>>, String> {
        let mut arguments = vec![];

        loop {
            let (arg_name_expr, arg_name) = parse_identifier(&mut self.lexer)?;
            self.lexer
                .expect(Kind::Op(Operator::Colon))
                .map_err(|e| e.to_string())?;
            let (arg_type, _) = parse_identifier(&mut self.lexer)?;

            arguments.push(Statement::FunArgument {
                name: arg_name,
                location: Location::new(arg_name_expr.location().start_byte, arg_type.location().end_byte),
                arg_type: Box::new(arg_type),
            });

            match self.lexer.peek().transpose().map_err(|e| e.to_string())? {
                Some(token) => match token.kind {
                    Kind::Op(Operator::RightParen) => break,
                    Kind::Op(Operator::Comma) => (),
                    _ => return Err(token.location.to_string())?,
                },
                None => break,
            }
        }

        Ok(arguments)
    }

    fn parse_function(&mut self) -> Result<Statement<'par>, String> {
        let keyword = self.lexer.expect(Kind::Fun).map_err(|e| e.to_string())?;
        let (_, fun_name) = parse_identifier(&mut self.lexer)?;
        self.lexer
            .expect(Kind::Op(Operator::LeftParen))
            .map_err(|e| e.to_string())?;

        let mut arguments = vec![];
        if let Some(next) = self.lexer.peek().transpose().map_err(|e| e.to_string())? {
            if !matches!(next.kind, Kind::Op(Operator::RightParen)) {
                arguments = self.parse_function_args(&keyword)?;
            }
        }

        // after parsing argument list we need to consume the closing parenthesis
        self.lexer
            .expect(Kind::Op(Operator::RightParen))
            .map_err(|e| e.to_string())?;

        // after the argument list of a function, there can be an optional return type annotation
        // => <TYPE> {
        // before the left brace, but its fine to be ommited
        let has_return = match self.lexer.peek().transpose().map_err(|e| e.to_string())? {
            Some(token) => matches!(token.kind, Kind::Op(Operator::ThickArrow)),
            _ => false,
        };

        let return_type = if has_return {
            self.lexer.next().transpose().map_err(|e| e.to_string())?;
            let (identifier, _) = parse_identifier(&mut self.lexer)?;
            Some(Box::new(identifier))
        } else {
            None
        };

        self.lexer
            .expect(Kind::Op(Operator::LeftBrace))
            .map_err(|e| e.to_string())?;

        let mut body = vec![];

        while let Some(token) = self.lexer.peek().transpose().map_err(|e| e.to_string())? {
            if let Kind::Op(Operator::RightBrace) = token.kind {
                break;
            }
            body.push(parse_expression(&mut self.lexer, true)?);
        }

        // consume the closing brace of the function
        let closing_brace = self
            .lexer
            .expect(Kind::Op(Operator::RightBrace))
            .map_err(|e| e.to_string())?;

        let location = keyword.location.start_byte..closing_brace.location.end_byte;
        Ok(Statement::Fun {
            name: fun_name,
            arguments,
            body,
            return_type,
            location: location.into(),
        })
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn make_sut(source: &str) -> Parser<'_> {
        let lexer = Lexer::new(source);
        Parser::new(source, lexer)
    }

    #[test]
    fn function_declaration() {
        let source = r#"
            fun calculate_circumference(diameter: f64) => f64 {
                const pi = 3.14159265358979323846264338327950288_f32;
                const radius = diameter / 2.0;
                const circumference = 2.0 * pi * radius;

                const nesting = {
                    const something = 10;
                    var nesting_more = {
                        return 10 + 3 * 4;
                    };
                    return 10 + something;
                };
            
                circumference
            }"#;

        let ast = match make_sut(source).parse() {
            Ok(expr) => expr,
            Err(e) => panic!("{e:?}"),
        };

        insta::assert_debug_snapshot!(ast);
    }

    #[test]
    fn language_features() {
        let source = r#"
            fun some_function_name(argument: TypeOfArg) => ReturnType {
                const immutable_var = if truthy_val {
                    return 10;
                } else if another_truthy == 10 {
                    return 20 + 3 * 2;
                } else {
                    const my_inner_var: f64 = 100; // comments don't matter
                    // this would be invalid when we do type checking
                    // as we never return!
                };

                /*
                 * we also have multiline comments! 
                    /* Although both regular
                     * and multiline comments will never appear on the parser
                     * ast, as they are ignored on the lexer 
                     **/
                 **/

                var mutable_value = function_call(immutable_var + 10, immutable_var);
                mutable_value = 10;

                {
                    const something = println(10 + 3);
                    println(10);
                    something_else();
                    var something = func_call();
                }

                mutable_value
            }
        "#;

        let ast = match make_sut(source).parse() {
            Ok(expr) => expr,
            Err(e) => panic!("{e:?}"),
        };

        insta::assert_debug_snapshot!(ast);
    }
}
