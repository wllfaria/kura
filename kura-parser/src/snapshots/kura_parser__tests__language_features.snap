---
source: kura-parser/src/lib.rs
expression: ast
---
[
    Fun {
        name: "some_function_name",
        arguments: [
            FunArgument {
                name: "argument",
                arg_type: Ident {
                    name: "TypeOfArg",
                    location: Location {
                        start_byte: 46,
                        end_byte: 55,
                    },
                },
                location: Location {
                    start_byte: 36,
                    end_byte: 55,
                },
            },
        ],
        body: [
            Var {
                mutable: false,
                name: "immutable_var",
                typ: None,
                value: If {
                    condition: Ident {
                        name: "truthy_val",
                        location: Location {
                            start_byte: 114,
                            end_byte: 124,
                        },
                    },
                    location: Location {
                        start_byte: 111,
                        end_byte: 477,
                    },
                    truthy: Block {
                        expressions: [
                            Return {
                                value: UintLiteral {
                                    value: 10,
                                    size: None,
                                    location: Location {
                                        start_byte: 154,
                                        end_byte: 156,
                                    },
                                },
                                location: Location {
                                    start_byte: 147,
                                    end_byte: 157,
                                },
                            },
                        ],
                        location: Location {
                            start_byte: 125,
                            end_byte: 175,
                        },
                    },
                    falsy: [
                        If {
                            condition: BinaryOp {
                                operator: EqualEqual,
                                lhs: Ident {
                                    name: "another_truthy",
                                    location: Location {
                                        start_byte: 184,
                                        end_byte: 198,
                                    },
                                },
                                rhs: UintLiteral {
                                    value: 10,
                                    size: None,
                                    location: Location {
                                        start_byte: 202,
                                        end_byte: 204,
                                    },
                                },
                                location: Location {
                                    start_byte: 184,
                                    end_byte: 204,
                                },
                            },
                            location: Location {
                                start_byte: 181,
                                end_byte: 477,
                            },
                            truthy: Block {
                                expressions: [
                                    Return {
                                        value: BinaryOp {
                                            operator: Plus,
                                            lhs: UintLiteral {
                                                value: 20,
                                                size: None,
                                                location: Location {
                                                    start_byte: 234,
                                                    end_byte: 236,
                                                },
                                            },
                                            rhs: BinaryOp {
                                                operator: Star,
                                                lhs: UintLiteral {
                                                    value: 3,
                                                    size: None,
                                                    location: Location {
                                                        start_byte: 239,
                                                        end_byte: 240,
                                                    },
                                                },
                                                rhs: UintLiteral {
                                                    value: 2,
                                                    size: None,
                                                    location: Location {
                                                        start_byte: 243,
                                                        end_byte: 244,
                                                    },
                                                },
                                                location: Location {
                                                    start_byte: 239,
                                                    end_byte: 244,
                                                },
                                            },
                                            location: Location {
                                                start_byte: 234,
                                                end_byte: 244,
                                            },
                                        },
                                        location: Location {
                                            start_byte: 227,
                                            end_byte: 245,
                                        },
                                    },
                                ],
                                location: Location {
                                    start_byte: 205,
                                    end_byte: 263,
                                },
                            },
                            falsy: [
                                Block {
                                    expressions: [
                                        Var {
                                            mutable: false,
                                            name: "my_inner_var",
                                            typ: Some(
                                                Ident {
                                                    name: "f64",
                                                    location: Location {
                                                        start_byte: 311,
                                                        end_byte: 314,
                                                    },
                                                },
                                            ),
                                            value: UintLiteral {
                                                value: 100,
                                                size: None,
                                                location: Location {
                                                    start_byte: 317,
                                                    end_byte: 320,
                                                },
                                            },
                                            location: Location {
                                                start_byte: 291,
                                                end_byte: 320,
                                            },
                                        },
                                    ],
                                    location: Location {
                                        start_byte: 269,
                                        end_byte: 477,
                                    },
                                },
                            ],
                        },
                    ],
                },
                location: Location {
                    start_byte: 89,
                    end_byte: 477,
                },
            },
            Var {
                mutable: true,
                name: "mutable_value",
                typ: None,
                value: FunCall {
                    ident: Ident {
                        name: "function_call",
                        location: Location {
                            start_byte: 820,
                            end_byte: 833,
                        },
                    },
                    location: Location {
                        start_byte: 820,
                        end_byte: 868,
                    },
                    arguments: [
                        BinaryOp {
                            operator: Plus,
                            lhs: Ident {
                                name: "immutable_var",
                                location: Location {
                                    start_byte: 834,
                                    end_byte: 847,
                                },
                            },
                            rhs: UintLiteral {
                                value: 10,
                                size: None,
                                location: Location {
                                    start_byte: 850,
                                    end_byte: 852,
                                },
                            },
                            location: Location {
                                start_byte: 834,
                                end_byte: 852,
                            },
                        },
                        Ident {
                            name: "immutable_var",
                            location: Location {
                                start_byte: 854,
                                end_byte: 867,
                            },
                        },
                    ],
                },
                location: Location {
                    start_byte: 800,
                    end_byte: 868,
                },
            },
            Assign {
                ident: Ident {
                    name: "mutable_value",
                    location: Location {
                        start_byte: 886,
                        end_byte: 899,
                    },
                },
                location: Location {
                    start_byte: 886,
                    end_byte: 905,
                },
                value: UintLiteral {
                    value: 10,
                    size: None,
                    location: Location {
                        start_byte: 902,
                        end_byte: 904,
                    },
                },
            },
            Block {
                expressions: [
                    Var {
                        mutable: false,
                        name: "something",
                        typ: None,
                        value: FunCall {
                            ident: Ident {
                                name: "println",
                                location: Location {
                                    start_byte: 963,
                                    end_byte: 970,
                                },
                            },
                            location: Location {
                                start_byte: 963,
                                end_byte: 978,
                            },
                            arguments: [
                                BinaryOp {
                                    operator: Plus,
                                    lhs: UintLiteral {
                                        value: 10,
                                        size: None,
                                        location: Location {
                                            start_byte: 971,
                                            end_byte: 973,
                                        },
                                    },
                                    rhs: UintLiteral {
                                        value: 3,
                                        size: None,
                                        location: Location {
                                            start_byte: 976,
                                            end_byte: 977,
                                        },
                                    },
                                    location: Location {
                                        start_byte: 971,
                                        end_byte: 977,
                                    },
                                },
                            ],
                        },
                        location: Location {
                            start_byte: 945,
                            end_byte: 978,
                        },
                    },
                    FunCall {
                        ident: Ident {
                            name: "println",
                            location: Location {
                                start_byte: 1000,
                                end_byte: 1007,
                            },
                        },
                        location: Location {
                            start_byte: 1000,
                            end_byte: 1011,
                        },
                        arguments: [
                            UintLiteral {
                                value: 10,
                                size: None,
                                location: Location {
                                    start_byte: 1008,
                                    end_byte: 1010,
                                },
                            },
                        ],
                    },
                    FunCall {
                        ident: Ident {
                            name: "something_else",
                            location: Location {
                                start_byte: 1033,
                                end_byte: 1047,
                            },
                        },
                        location: Location {
                            start_byte: 1033,
                            end_byte: 1049,
                        },
                        arguments: [],
                    },
                    Var {
                        mutable: true,
                        name: "something",
                        typ: None,
                        value: FunCall {
                            ident: Ident {
                                name: "func_call",
                                location: Location {
                                    start_byte: 1087,
                                    end_byte: 1096,
                                },
                            },
                            location: Location {
                                start_byte: 1087,
                                end_byte: 1098,
                            },
                            arguments: [],
                        },
                        location: Location {
                            start_byte: 1071,
                            end_byte: 1098,
                        },
                    },
                ],
                location: Location {
                    start_byte: 923,
                    end_byte: 1117,
                },
            },
            Ident {
                name: "mutable_value",
                location: Location {
                    start_byte: 1135,
                    end_byte: 1148,
                },
            },
        ],
        return_type: Some(
            Ident {
                name: "ReturnType",
                location: Location {
                    start_byte: 60,
                    end_byte: 70,
                },
            },
        ),
        location: Location {
            start_byte: 13,
            end_byte: 1162,
        },
    },
]
